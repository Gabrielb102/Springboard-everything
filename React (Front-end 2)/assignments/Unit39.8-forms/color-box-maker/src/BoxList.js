import Box from "./Box";
import Title from "./Title";
import NewBoxForm from "./NewBoxForm";
import {useState} from "react";
import { v4 as uuid } from 'uuid';

const BoxList = () => {
    const [boxes, setBoxes] = useState([])

    const removeFunc = (id) => {
        
        const boxToDelete = boxes.find(box => {
            box.props.id === id});
        const idx = boxes.indexOf(boxToDelete);
        setBoxes(boxes => {
            boxes = boxes.splice(idx, 1);
            return boxes;
        });
    }

    const addBoxFunc = (color, height, width) => {
        const id = uuid();
        const newBox = <Box id={id} color={color} height={height} width={width} removeFunc={() => removeFunc(id)}/>;
        setBoxes(boxes => boxes = [...boxes, newBox]);
    }

    return (
        <>
            <Title length={boxes.length}/>
            <NewBoxForm addBoxFunc={addBoxFunc}/>
            {boxes}
        </>
    )}

export default BoxList;