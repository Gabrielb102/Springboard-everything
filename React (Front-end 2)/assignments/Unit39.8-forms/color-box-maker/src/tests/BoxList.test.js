import { render, fireEvent, getByTestId } from "@testing-library/react";
import BoxList from "../BoxList";

describe("Tests for BoxList Component", () => {

    it("Renders", () => {
        render(<BoxList/>);
    })

    it("Renders a consistent user interface (no boxes)", () => {
        const { asFragment } = render(<BoxList />);
        expect(asFragment).toMatchSnapshot();
    })

    it("Adds a box when form is submitted", () => {
        const { getByText, getByTestId, getByLabelText } = render(<BoxList/>);
        const submitButton = getByText("New Box");
        const colorInput = getByLabelText("Color");
        const heightInput = getByLabelText("Height");
        const widthInput = getByLabelText("Width");
        fireEvent.change(colorInput, {target: {value: "red"}});
        fireEvent.change(heightInput, {target: {value: "10px"}});
        fireEvent.change(widthInput, {target: {value: "100px"}});
        fireEvent.click(submitButton);
        const box = getByTestId("box")
        expect(box).not.toEqual(null);
    })

    it("appears when form is submitted", () => {
        const handleSubmit = jest.fn((e) => {
            e.preventDefault()
            addBoxFunc("red", "250px", "50px");
            setFormData(INITIALCONDITIONS);
        })
        const {getByTestId} = render(<BoxList/>);
        const testForm = getByTestId("form");
        fireEvent.submit(testForm);
        const testBoxes = getByTestId("box");
        expect(testBoxes).toBeInDocument;

    })

    it("Removes box when the x is clicked", () => {
        const { getByText, queryByTestId, getByLabelText } = render(<BoxList/>);
        const submitButton = getByText("New Box");
        const colorInput = getByLabelText("Color");
        const heightInput = getByLabelText("Height");
        const widthInput = getByLabelText("Width");
        fireEvent.change(colorInput, {target: {value: "red"}});
        fireEvent.change(heightInput, {target: {value: "10px"}});
        fireEvent.change(widthInput, {target: {value: "100px"}});
        fireEvent.click(submitButton);
        const testBoxesBeforeX = queryByTestId('box');
        const xButton = getByText("x");
        fireEvent.click(xButton);
        // Cannot figure out debugger, turn in and follow up with mentor
        const testBoxesAfterX = queryByTestId('box');
        expect(testBoxesBeforeX).not.toEqual(null);
        expect(testBoxesAfterX).toEqual(null);
    })


})