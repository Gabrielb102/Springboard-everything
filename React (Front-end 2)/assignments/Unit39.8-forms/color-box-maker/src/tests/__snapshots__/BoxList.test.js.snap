// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for BoxList Component Renders a consistent user interface (no boxes) 1`] = `[Function]`;
